app/scripts/controllers/detect-tokens.js

```diff
@@ -194,6 +198,30 @@ export default class DetectTokensController {
     this.interval = DEFAULT_INTERVAL;
   }
 
+  /**
+   * Restart/Reschedule token detection polling period and call detectNewTokens
+   * in case of address change or user session initialization.
+   *
+   */
+  /* eslint-disable no-undef */
+  rescheduleTokenDetectionPollingInMV3() {
+    if (!(this.isActive && this.selectedAddress)) {
+      return;
+    }
+    chrome.alarms.clear(DETECT_TOKEN_ALARM);
+    chrome.alarms.create(DETECT_TOKEN_ALARM, {
+      delayInMinutes: DEFAULT_INTERVAL_MV3,
+      periodInMinutes: DEFAULT_INTERVAL_MV3,
+    });
+
+    chrome.alarms.onAlarm.addListener(() => {
+      const alarm = chrome.alarms.get(DETECT_TOKEN_ALARM);
+      if (alarm) {
+        this.detectNewTokens();
+      }
+    });
+  }
+
```

Comment:
We should use alarm only in scenario where we need code to keep running even after extension is dead as alarms are remembered by the browser

Reply:
**> +      return;
+    }
+    chrome.alarms.clear(DETECT_TOKEN_ALARM);
+    chrome.alarms.create(DETECT_TOKEN_ALARM, {
+      delayInMinutes: DEFAULT_INTERVAL_MV3,
+      periodInMinutes: DEFAULT_INTERVAL_MV3,
+    });
+
+    chrome.alarms.onAlarm.addListener(() => {
+      const alarm = chrome.alarms.get(DETECT_TOKEN_ALARM);
+      if (alarm) {
+        this.detectNewTokens();
+      }
+    });
+  }
+**

Reply:
detect-token needs to be called only when the extension is open and unlocked. Does that mean we do not need to add chrome alarms for this controller?

Reply:
Yes I think so, lets get more opinions on it.

Reply:
I will create a post on the MV3 channel regarding this concern

Reply:
I think it makes sense in this case so that we don't increase the number of requests that we are making to tokens api, it also ensures that the tokens list gets updated regularly, and the service worker restart doesn't accidently cause missed updates to the list

Reply:
I am not sure if the same controller context will be available when the controller reloads.

Reply:
>I am not sure if the same controller context will be available when the controller reloads.

@jpuri do you mean that the `this` reference could be incorrect?

Reply:
yes, when service worker reloads a new instance of controller will be created

Reply:
Is this conversation resolved @jpuri / @NiranjanaBinoy ?

Reply:
@danjm is helping me to dig deeper into how the change in context during restart is affecting our results; during the initial testings, it seems to be generating correct results but confirming that.

Reply:
Is this resolved yet?

Reply:
This is not resolved yet. If we are storing the results in scope of the current alarm function it will not be useful when controller is re-initiated.

If data if very important we can check to store it in storage.

Also this can result in multiple alarms being initiated each time new instance of controller is created and they may continue to run in browser.